HW0/P4/P4.py

```diff
@@ -40,38 +59,115 @@
 
     #####################
     # Part 3:
     # Use the initial conditions and propagation matrix for prediction
     #####################
 
-    # A = ?
-    # a = ?
-    # s = ?
+    # Initialize progation matrix
+    A = np.mat([
+        [1, 0, 0, dt, 0, 0],
+        [0, 1, 0, 0, dt, 0],
+        [0, 0, 1, 0, 0, dt],
+        [0, 0, 0, 1-c*dt, 0, 0],
+        [0, 0, 0, 0, 1-c*dt, 0],
+        [0, 0, 0, 0, 0, 1-c*dt]
+        ])
+    
+    # Initialize a matrix (impact of gravity)
+    a = np.mat([
+        [0],
+        [0],
+        [0],
+        [0],
+        [0],
+        [g*dt]
+        ])
+    
+    # Construct 6xK matrix to fill in with predicted trajecory
+    s = np.mat(np.zeros((6, K)))
 
     # Initial conditions for s0
+    s[:,0] = np.mat(s_true[0,:]).transpose()
+    
     # Compute the rest of sk using Eq (1)
+    for k in range(1,K):
+        s[:,k] = (A * s[:,k-1]) + a
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '-k', label='Blind trajectory')
+    # Convert matrix to array for plotting purposes
+    s = np.array(s)
+        
+    # Plot estimated trajectory (simple model)
+    ax.plot(s[0,:], s[1,:], s[2,:],
+            '-k', label='Blind trajectory')
 
     #####################
     # Part 4:
     # Use the Kalman filter for prediction
     #####################
+    
+    # Initialize B matrix
+    B = np.mat([
+        [bx, 0, 0, 0, 0, 0],
+        [0, by, 0, 0, 0, 0],
+        [0, 0, bz, 0, 0, 0],
+        [0, 0, 0, bvx, 0, 0],
+        [0, 0, 0, 0, bvy, 0],
+        [0, 0, 0, 0, 0, bvz]
+        ])
+    
+    # Initialize C matrix
+    C = np.mat([
+        [rx, 0, 0, 0, 0, 0],
+        [0, ry, 0, 0, 0, 0],
+        [0, 0, rz, 0, 0, 0]
+        ])
+
+    # Construct 6xK matrix to fill in with predicted trajecory
+    s_kal = np.mat(np.zeros((6, K)))
 
-    # B = ?
-    # C = ?
-
-    # Initial conditions for s0 and Sigma0
+    # Initial conditions for s0
+    s_kal[:,0] = np.mat(s_true[0,:]).transpose()
+    
+    # Initial conditions for sigma0
+    sigma_init = 0.01
+    sigma = np.mat(np.eye(6)) * sigma_init
+    
+    # Convert array to matrix to simplify calculations
+    s_obs = np.mat(s_obs)
+                                                  
     # Compute the rest of sk using Eqs (2), (3), (4), and (5)
-
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '-r', label='Filtered trajectory')
+    # Functions defined above
+    for k in range(1,K):
+        s_kal[:,k] = predictS(s_kal[:,k-1])
+        sigma_pred = predictSig(sigma)
+        sigma = updateSig(sigma_pred)
+        s_kal[:,k] = updateS(s_kal[:,k], s_obs[k,:].transpose(), sigma_pred, sigma)
+    
+    # Convert matrix to array for plotting purposes
+    s_kal = np.array(s_kal)
+    
+    # Plot filtered trajectory
+    ax.plot(s_kal[0,:], s_kal[1,:], s_kal[2,:],
+            '-r', label='Filtered trajectory')
 
     # Show the plot
     ax.legend()
-    plt.show()
+    plt.show()
```

Comment:
Unable to comment on the graph itself, so I will write it here. but I think it is better to have both the lone cashier and infinite cashier having the same y-axis values. The graph below gives the first impression that the infinite cashier takes a longer time than the lone cashier